pkg/render/csi.go

```diff
@@ -0,0 +1,334 @@
+
+	csiImage          string
+	csiRegistrarImage string
+}
+
+func CSI(cfg *CSIConfiguration) Component {
+	return &csiComponent{
+		cfg: cfg,
+	}
+}
+
+func (c *csiComponent) csiDriver() *v1.CSIDriver {
+	meta := metav1.ObjectMeta{
+		Name: CSIDriverName,
+	}
+
+	typeMeta := metav1.TypeMeta{
+		Kind:       "CSIDriver",
+		APIVersion: "storage/v1",
+	}
+
+	volumeLifecycleModes := []v1.VolumeLifecycleMode{
+		v1.VolumeLifecycleEphemeral,
+	}
+	spec := v1.CSIDriverSpec{
+		PodInfoOnMount:       ptr.BoolToPtr(true),
+		VolumeLifecycleModes: volumeLifecycleModes,
+	}
+
+	return &v1.CSIDriver{
+		TypeMeta:   typeMeta,
+		ObjectMeta: meta,
+		Spec:       spec,
+	}
+}
+
+func (c *csiComponent) csiTolerations() []corev1.Toleration {
+	operator := corev1.TolerationOperator(CSITolerationOperator)
+	tolerations := []corev1.Toleration{
+		corev1.Toleration{
+			Key:      CSITolerationControlPlaneKey,
+			Operator: operator,
+			Effect:   corev1.TaintEffectNoSchedule,
+		},
+		corev1.Toleration{
+			Key:      CSITolerationMasterKey,
+			Operator: operator,
+			Effect:   corev1.TaintEffectNoSchedule,
+		},
+	}
+	return tolerations
+}
+
+func (c *csiComponent) csiContainers() []corev1.Container {
+	mountPropagation := corev1.MountPropagationBidirectional
+	csiContainer := corev1.Container{
+		Name:            CSIContainerName,
+		Image:           c.csiImage,
+		ImagePullPolicy: corev1.PullAlways,
+		Args: []string{
+			"--nodeid=$(KUBE_NODE_NAME)",
+			"--loglevel=$(LOG_LEVEL)",
+		},
+		SecurityContext: &corev1.SecurityContext{
+			Privileged: ptr.BoolToPtr(true),
+		},
+		Env: []corev1.EnvVar{
+			corev1.EnvVar{
+				Name:  "LOG_LEVEL",
+				Value: "warn",
+			},
+			corev1.EnvVar{
+				Name: "KUBE_NODE_NAME",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						FieldPath: "spec.nodeName"},
+				},
+			},
+		},
+		VolumeMounts: []corev1.VolumeMount{
+			corev1.VolumeMount{
+				Name:      "varrun",
+				MountPath: "/var/run",
+			},
+			corev1.VolumeMount{
+				Name:      "etccalico",
+				MountPath: "/etc/calico",
+			},
+			corev1.VolumeMount{
+				Name:      "socket-dir",
+				MountPath: "/csi",
+			},
+			corev1.VolumeMount{
+				Name:             "kubelet-dir",
+				MountPath:        "/var/lib/kubelet",
+				MountPropagation: &mountPropagation,
+			},
+		},
+	}
+
+	// Construct "csi-node-driver-registrar" container
+	registrarContainer := corev1.Container{
+		Name:            CSIRegistrarContainerName,
+		Image:           c.csiRegistrarImage,
+		ImagePullPolicy: corev1.PullAlways,
+		Args: []string{
+			"--v=5",
+			"--csi-address=$(ADDRESS)",
+			"--kubelet-registration-path=$(DRIVER_REG_SOCK_PATH)",
+		},
+		SecurityContext: &corev1.SecurityContext{
+			Privileged: ptr.BoolToPtr(true),
+		},
+		Env: []corev1.EnvVar{
+			corev1.EnvVar{
+				Name:  "ADDRESS",
+				Value: "/csi/csi.sock",
+			},
+			corev1.EnvVar{
+				Name:  "DRIVER_REG_SOCK_PATH",
+				Value: "/var/lib/kubelet/plugins/csi.tigera.io/csi.sock",
```

Comment:
I'm wondering if we should be exposing this or any of the other paths that are below. Should we change VolumePlugin to be a struct that way we can expose any of these paths in the future if we need to without adding other fields at the same level?

Reply:
Do you mean rather than VolumePlugin simply being "Enabled" or "Disabled" at the API level the user can specify a series of kv pairs that would then be populated here?

Reply:
No we shouldn't simply expose kv pairs but I was thinking we would create a field for each path that the user would need to specify and those fields would be in an embedded in a struct so we might have something like:
```
spec:
  volumePlugin:
    sockPath: ...
    kubeletPath: ...
    registrationDir: ...
```
I think we'd also need a field to set Enable/Disable but I'm not sure what the name of the field should be.

Reply:
Ok I see what you mean, let me come up with a draft of this.